apps/src/code-studio/progress.js

```diff
@@ -288,8 +288,10 @@ function initializeStoreWithProgress(
   // Determine if we are viewing student progress.
   var isViewingStudentAnswer = !!clientState.queryParams('user_id');
 
-  // Merge in progress saved on the client, unless we are viewing student's work.
-  if (!isViewingStudentAnswer) {
+  // Merge in progress saved on the client, unless we are viewing student
+  // work or are logged in.
+  const useDbProgress = store.getState().progress.shouldUseDbProgress;
```

Comment:
two little NITs here:

1) the comment doesn't match the code. if `progress.shouldUseDbProgress` actually means "user is logged in", you should rename it to reflect that (and probably trigger the action from code that verifies the logged-in status instead of the presence of progress from the server). otherwise maybe just update the comment -- "viewing student's work or have server progress" or something?

2) since we clear client progress when setting `shouldUseDbProgress = true`, merging the client state would be a no-op in that case, right? this code would be slightly easier to read if it were just about handling the case of viewing a student's work.

Reply:
Ok. Had to put a good bit of thought into this: 
1. You exposed an issue with this change here. `shouldUseDbProgress` _is_ equivalent to "user is logged in." Thinking back, I  think I set it up this way because there are places in the code where the client currently doesn't have access to the logged in state of the user. I think some of the places that set shouldUseDbProgress will have have gotten a response with progress from the server, but the logged-in state will not yet have been set in the redux store from the session cookie. I'm going to look into whether that's the case for setting this value. If it is, that seems like some tech debt that should be cleaned up. Logged-in state seems like something that should always be accessible. 
2. That is true. This area of the code is still a little fragile though. My goal was to completely separate the DB progress from the SessionStorage (clientState) progress. So if we get progress from the DB (i.e. the user is logged in), we should never touch the SessionStorage progress. And if we get no progress from the DB (i.e. the user is logged out), we should only touch the SessionStorage progress. I'm hoping by using this check here, we make that distinction more explicit. I'm curious for your thoughts though - this has a bit of an odd setup, and I want to improve it rather than make it unclear.

Reply:
i mentioned these things as NITs because i figured they were likely related to tech debt that is out of scope of this feature. i think auditing and standardizing the setting and availability of logged-in state is probably a worthwhile undertaking but should be its own task, which i imagine won't be trivial.

re 1)
one thought, and this is definitely a NIT, is that renaming `shouldUseDbProgress` to `usingDbProgress` might make this all a little clearer -- i think `usingDbProgress` is easier to intuitively connect to "is logged in", whereas `shouldUseDbProgress` immediately begs the question of "why?" (for someone like me unfamiliar with the code), which makes my brain think there's something more complicated i need to understand. so that could be a compromise that doesn't require making sure that this value always means "is logged in".

could even take it a step farther and call it `usingLoggedInDbProgress` if you want to make it really clear.

re 2)
i figured fragility might be the case here, and robustness is a fine excuse for slight reduction in clarity. i do think though that renaming this value to `usingDbProgress` (regardless of whether you rename the property in the redux store), combined with breaking up this compound conditional, will make this much easier to understand. `useDbProgress` combined with the fact that the negation is far away from the value made me have to work out whether this meant that if we're using db progress we should `mergeProgress`. `!usingDbProgress` would make it much clearer that we don't enter this block when using the db. but again, this is a NIT!

Reply:
These are great suggestions! Really good questions too. It forced me to put some significant additional thought into my approach (which is a good thing)! Definitely wins the award of the most useful (and thought-provoking) NIT of the year.

Reply:
hah! :)

great additions of comments around this!